File: py/prim.py
import logging
from typing import Optional

from util import Edge, Network, Vertice

logger = logging.getLogger('prim')


def find_cheaper_edge(visited_vertices: list[Vertice], edges: list[Edge]) -> Optional[Edge]:
    log = logger.getChild(find_cheaper_edge.__name__)
    result = None
    for edge in edges:
        if edge.u in visited_vertices and edge.v not in visited_vertices:
            log.debug(f'potential link: {edge}, current: {result}')
            if result is None or result.cost > edge.cost:
                result = edge

    return result


def solve(network: Network) -> Network:
    log = logger.getChild(solve.__name__)

    edges = network.edges_sorted()
    vertices = network.vertices_sorted()
    vertices_count = len(vertices)

    log.debug(f'vertices_count={vertices_count}, vertices={vertices}')
    log.debug(f'edges_count={len(edges)}, edges={edges}')

    tree = Network()

    visited_vertices = [vertices[0]]  # include one vertice to start from

    for step in range(vertices_count - 1):
        log.debug(f'step {step}, visited_vertices {visited_vertices}')
        edge = find_cheaper_edge(visited_vertices, edges)
        if edge is not None:
            visited_vertices.append(edge.v)
            tree.insert_edge(edge)

    return tree

File: src/prim.rs
use std::collections::HashSet;

use crate::solv::{Edge, Network, Solver, VerticeId};

pub struct Prim {}

impl Prim {
    fn find_cheaper_edge(visisted_vertices: &HashSet<VerticeId>, edges: &[Edge]) -> Option<Edge> {
        let mut result: Option<Edge> = None;
        let vertice_is_present = |vertice: &VerticeId| visisted_vertices.get(vertice).is_some();
        let vertice_is_absent = |vertice: &VerticeId| visisted_vertices.get(vertice).is_none();

        for edge in edges {
            if vertice_is_present(&edge.u) && vertice_is_absent(&edge.v) {
                log::debug!("potential_edge={}, current_edge={:?}", edge, result);

                if let Some(ref current_edge) = result {
                    if current_edge.weight > edge.weight {
                        result.replace(edge.clone());
                    }
                } else {
                    result.replace(edge.clone());
                }
            }
        }

        result
    }
}

impl Solver for Prim {
    fn solve(network: Network) -> Network {
        let mut tree = Network::default();
        let vertices = {
            let mut v = network.vertices();
            v.sort();
            v
        };
        let edges = network.edges();
        let mut visisted_vertices = {
            let mut set = HashSet::default();
            set.insert(vertices[0].clone());
            set
        };

        log::debug!("vertices_count={}, vertices={:?}", vertices.len(), vertices);
        log::debug!("edges_count={}, edges={:?}", edges.len(), edges);

        for step in 0..vertices.len() {
            log::debug!("step={}, visisted_vertices={:?}", step, visisted_vertices);
            if let Some(edge) = Prim::find_cheaper_edge(&visisted_vertices, &edges) {
                tree.insert_edge(edge.clone());
                visisted_vertices.insert(edge.v);
            }
        }

        tree
    }
}
